Stage_4

KinCadeFoster opened the pull request:
- Реализовано для режима изучения слов отображение загаданного и четыре перемешанных невыученных слова.
- Для случая когда неизученных слов меньше чем 4, берем слова из общего списка с изыченными словами
- Добавлена проверка на невыученные слова. Если таких нет, отображать “Вы выучили все слова”.

KinCadeFoster committed:
Stage_4
- Реализовано для режима изучения слов отображение загаданного и четыре перемешанных невыученных слова.
- Для случая когда неизученных слов меньше чем 4, берем слова из общего списка с изыченными словами
- Добавлена проверка на невыученные слова. Если таких нет, отображать “Вы выучили все слова”.

danelyan commented on src/main/kotlin/Main.kt:
```diff
@@ -34,21 +34,92 @@ fun getDictionary(): List<Word> {
+}
+
+fun getUnlearnedWords(dictionary: List<Word>): List<Word> {
+    return dictionary.filter { it.correctCount < 3 }
```
3 можно вынести в константы

KinCadeFoster replied:
добавил константу

danelyan commented on src/main/kotlin/Main.kt:
```diff
@@ -34,21 +34,92 @@ fun getDictionary(): List<Word> {
+    return dictionary.filter { it.correctCount < 3 }
+}
+
+fun updateCorrectCount(dictionary: List<Word>, learningWord: List<Word>) {
```
этот метод излишен. У тебя уже есть слово которое является правильным. Подробнее выше отписал

KinCadeFoster replied:
удалил метод

danelyan commented on src/main/kotlin/Main.kt:
```diff
@@ -3,7 +3,7 @@ import java.io.File
+    var correctCount: Int,
 )
 
 fun getUserStatistics(dictionary: List<Word>) {
```
Давай 2 вынесем в константы с нормальным именем.
Точнее, вынесем 3 и воспользуемся >=

KinCadeFoster replied:
добавил константу

danelyan commented on src/main/kotlin/Main.kt:
```diff
@@ -34,21 +34,92 @@ fun getDictionary(): List<Word> {
+        print("Введите ваш ответ: ")
+        val userAnswer = readlnOrNull()?.toInt() ?: 404
+        if (userAnswer == 0) break
+        if (userAnswer !in 1..4) {
```
а если вариантов больше 4? Давай вынесем в константу

KinCadeFoster replied:
добавил константу

danelyan commented on src/main/kotlin/Main.kt:
```diff
@@ -34,21 +34,92 @@ fun getDictionary(): List<Word> {
+        val learningWord = takeShuffleElements(unLearningWord = unlearnedWords)
+        println("Перевод слова <${learningWord[0].enWord.lowercase().replaceFirstChar { it.uppercase() }}>")
+
+        val answers = takeShuffleElements(dictionary, unlearnedWords, value = 3, learningWord)
```
3 - в константы

KinCadeFoster replied:
добавил константу

danelyan commented on src/main/kotlin/Main.kt:
```diff
@@ -34,21 +34,92 @@ fun getDictionary(): List<Word> {
+        unLearningWord.shuffled().take(1)
+    } else if (unLearningWord.size - 1 < value) {
+        val tempDictionary = dictionary.filter { word -> !learningWord.contains(word) }
+        val answers = tempDictionary.shuffled().take(value) + learningWord
```
тут много повторяющегося кода.
Можем его обобщить?
К примеру 83, 84 и 85 строка идентичны 87, 88, 89

KinCadeFoster replied:
добавил метод getAnswerChoices

KinCadeFoster committed:
Stage_4
- вынес константы WORD_LEARNED_LIMIT, TOTAL_ANSWER_CHOICES, TOTAL_ANSWER_CHOICES
- удалил метод updateCorrectCount, логику упростил, перенес в основной код
- добавил метод getAnswerChoices, теперь код не дублируется

danelyan reviewed (approved):
Хорошо. Дальше еще немного порефакторим )
